src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -256,11 +435,97 @@ public Void visitMethod(MethodTree tree, AnnotatedTypeMirror p) {
       ExecutableElement element = TreeUtils.elementFromDeclaration(tree);
       if (ElementUtils.hasAnnotation(element, Provides.class.getName())) {
         // TODO: Put fully qualified names of classes in knownBindings
-        knownBindings.put(tree.getReturnType().toString(), p.getUnderlyingType().toString());
+        knownBindings.put(
+            tree.getReturnType().toString(),
+            new KnownBindingsValue.KnownBindingsValueBuilder(p.getUnderlyingType().toString())
+                .build());
       }
 
       printKnownBindings();
       return super.visitMethod(tree, p);
     }
   }
+
+  @Override
+  protected QualifierHierarchy createQualifierHierarchy() {
+    return new DependencyInjectionQualifierHierarchy(
+        this.getSupportedTypeQualifiers(), this.elements);
+  }
+
+  protected class DependencyInjectionQualifierHierarchy extends AccumulationQualifierHierarchy {
+
+    private final AnnotationMirror bawAnno =
+        AnnotationBuilder.fromClass(elements, BindAnnotatedWith.class);
+
+    protected DependencyInjectionQualifierHierarchy(
+        Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+      super(qualifierClasses, elements);
+    }
+
+    @Override
+    public boolean isSubtype(final AnnotationMirror subAnno, final AnnotationMirror superAnno) {
+
+      if (AnnotationUtils.areSameByName(subAnno, bawAnno)
+          && AnnotationUtils.areSameByName(superAnno, bawAnno)) {
+        if (AnnotationUtils.areSame(subAnno, superAnno)) {
+          return true;
+        }
+        return false;
+      }
+
+      if (AnnotationUtils.areSameByName(subAnno, bawAnno)
+          || AnnotationUtils.areSameByName(superAnno, bawAnno)) {
+        return false;
+      }
+
+      return super.isSubtype(subAnno, superAnno);
+    }
+
+    @Override
+    public @Nullable AnnotationMirror leastUpperBound(
+        final AnnotationMirror qualifier1, final AnnotationMirror qualifier2) {
+
+      if (AnnotationUtils.areSame(qualifier1, qualifier2)) {
+        return qualifier1;
+      }
+
+      if (AnnotationUtils.areSameByName(qualifier1, bawAnno)
+          || AnnotationUtils.areSameByName(qualifier2, bawAnno)) {
+        return top;
+      }
+
+      return super.leastUpperBound(qualifier1, qualifier2);
+    }
+
+    @Override
+    public @Nullable AnnotationMirror greatestLowerBound(
+        AnnotationMirror qualifier1, AnnotationMirror qualifier2) {
+
+      if (AnnotationUtils.areSame(qualifier1, qualifier2)) {
+        return qualifier1;
+      }
+
+      if (AnnotationUtils.areSameByName(qualifier1, bawAnno)
+          || AnnotationUtils.areSameByName(qualifier2, bawAnno)) {
+        return bottom;
+      }
+
+      return super.greatestLowerBound(qualifier1, qualifier2);
+    }
+  }
+
+  public AnnotationMirror createBindAnnotatedWithAnnotation(
+      List<String> values, List<String> names) {
```

Comment:
Shouldn't `@BindAnnotatedWith` only ever have a single `value` element? If so, it doesn't make sense to provide a factory method that takes a `List`.

Reply:
I think I created it like this based on the patterns I've seen implemented for `@Bind`. I definitely do think that `@BAW` will only ever have a single `value` element as well as a single `annotatedWith` element. I think this is something I should change in an upcoming PR.

Reply:
>  I definitely do think that @BAW will only ever have a single value element as well as a single annotatedWith element

In that case, we should just delete this version of this method, and only let callers use the version that takes single elements. There's no need to wait for a future PR.

Reply:
Done.